src/Catalyst.KBucket/KBucket.cs

```diff
@@ -0,0 +1,528 @@
+        /// </value>
+        public int ContactsPerBucket { get; set; } = 20;
+
+        /// <summary>
+        ///   The number of contacts to ping when a bucket that should not be split
+        ///   becomes full.
+        /// </summary>
+        /// <value>
+        ///   Defaults to 3.
+        /// </value>
+        /// <seealso cref="Ping"/>
+        public int ContactsToPing { get; set; } = 3;
+
+        /// <summary>
+        ///   The ID of the local contact/peer.
+        /// </summary>
+        /// <value>
+        ///   Defaults to 20 random bytes.
+        /// </value>
+        public byte[] LocalContactId
+        {
+            get
+            {
+                if (_localContactId != null) return _localContactId;
+
+                _localContactId = new byte[20];
+                new Random().NextBytes(_localContactId);
+
+                return _localContactId;
+            }
+            set => _localContactId = value;
+        }
+
+        /// <summary>
+        ///   The root of the binary tree.
+        /// </summary>
+        public Bucket<T> Root { get; private set; } = new Bucket<T>();
+
+        /// <summary>
+        ///   Raised when a bucket needs splitting but cannot be split.
+        /// </summary>
+        public event EventHandler<PingEventArgs<T>> Ping;
+
+        /// <summary>
+        ///   Determines which contact is used when both have the same ID.
+        /// </summary>
+        /// <value>
+        ///   Defaults to <see cref="DefaultAbiter(T, T)"/>.
+        /// </value>
+        /// <remarks>
+        ///   The arguments are the incumbent and the candidate.
+        /// </remarks>
+        public Func<T, T, T> Arbiter { get; set; } = DefaultAbiter;
+
+        /// <summary>
+        ///   Used to determine which contact should be use when both
+        ///   have the same ID.
+        /// </summary>
+        /// <param name="incumbent">
+        ///   The existing contact.
+        /// </param>
+        /// <param name="candidate">
+        ///   The new contact.
+        /// </param>
+        /// <returns>
+        ///   Always returns the <paramref name="incumbent"/>.
+        /// </returns>
+        public static T DefaultAbiter(T incumbent, T candidate) { return incumbent; }
+
+        /// <summary>
+        ///   Finds the XOR distance between the two contacts.
+        /// </summary>
+        public UInt256 Distance(T a, T b)
+        {
+            Validate(a);
+            Validate(b);
+            return Distance(a.Id, b.Id);
+        }
+
+        /// <summary>
+        ///   Finds the XOR distance between the two contact IDs.
+        /// </summary>
+        public UInt256 Distance(byte[] a, byte[] b)
+        {
+            UInt256 distance = 0;
+            var i = 0;
+            var min = Math.Min(a.Length, b.Length);
+            var max = Math.Max(a.Length, b.Length);
+            for (; i < min; ++i) distance = distance * 256 + ((ulong) a[i] ^ b[i]);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'for' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)